sources/airtable/__init__.py

```diff
@@ -0,0 +1,68 @@
+"""Source that loads tables form Airtable.
+Supports whitelisting of tables or loading of all tables from a specified base.
+Supported write dispositions: replace
+"""
+from typing import Optional, Iterable, Iterator, List, Any
+
+import dlt
+from dlt.extract.source import DltResource
+from dlt.common.typing import TDataItem
+
+import pyairtable
+
+
+@dlt.source
+def airtable_source(
+    base_id: str,
+    table_ids: Optional[List[str]] = None,
+    table_names: Optional[List[str]] = None,
+    access_token: str = dlt.secrets.value,
+) -> Iterable[DltResource]:
+    """
+    Represents tables for a single Airtable base.
+    Args:
+        base_id (str): The id of the base. Obtain it e.g. from the URL in your webbrowser. It starts with "app". See https://support.airtable.com/docs/finding-airtable-ids
+        table_ids (Optional[List[str]]): A list of table ids to load. By default, all tables in the schema are loaded. Starts with "tbl". See https://support.airtable.com/docs/finding-airtable-ids
+        table_names (Optional[List[str]]): A list of table names to load. By default, all tables in the schema are loaded. Table names can change and thus filtering on names is less reliable than on ids.
+        access_token (str): The personal access token. See https://support.airtable.com/docs/creating-and-using-api-keys-and-access-tokens#personal-access-tokens-basic-actions
+    """
+    api = pyairtable.Api(access_token)
+    all_tables_url = api.build_url(f"meta/bases/{base_id}/tables")
+    tables = api.request(method="GET", url=all_tables_url).get("tables")
+    for t in tables:
+        if table_ids:
+            if t.get("id") in table_ids:
+                yield airtable_resource(access_token, base_id, t)
+        elif table_names:
+            if t.get("name") in table_names:
+                yield airtable_resource(access_token, base_id, t)
+        else:
+            yield airtable_resource(access_token, base_id, t)
+
+
+def airtable_resource(
+    access_token: str,
+    base_id: str,
+    table: TDataItem,
+) -> DltResource:
+    """
+    Represents a single airtable.
+    Args:
+        base_id (str): The id of the base. Obtain it e.g. from the URL in your webbrowser. It starts with "app". See https://support.airtable.com/docs/finding-airtable-ids
+        access_token (str): The personal access token. See https://support.airtable.com/docs/creating-and-using-api-keys-and-access-tokens#personal-access-tokens-basic-actions
+        table (TDataItem): The iterable created by pyairtable representing the data of an airtable
+    """
+    primary_key_id = table["primaryFieldId"]
+    primary_key_field = [
+        field for field in table["fields"] if field["id"] == primary_key_id
+    ][0]
+    table_name: str = table["name"]
+    primary_key: List[str] = [primary_key_field["name"]]
+    air_table = pyairtable.Table(access_token, base_id, table["id"])
```

Comment:
here's one small optimization you can do. since not always all tables will be opened (people can select just a few resources from source with ) let's create a small inner generator function:
```
def air_table_generator():
   air_table = pyairtable.Table(access_token, base_id, table["id"])
   yield from air_table.iterate()  # here we could chunk. I wonder if Aritable supports chunking?
   
return dlt.resource(
        air_table_generator,  # don't call!
```

tell me if that worked

Reply:
Yes, the Airtable API wrapper library fits well with dlt because it does chunking. 

Thank you for the hint towards the generator. Will this increase efficiency even if there are mostly less than a dozen tables in one base? I ask because I've seen benchmarks of Python list vs. generator which suggested that the setup cost is quite high and thus for short lists a generator is much slower – not sure if that "slower" would be noticeable though :)

Reply:
I would prefer to not have an internal generator for generators because:

1. the dlt resource is already created with a generator because the call to `airtable.iterate()` already returns a generator.
2. the function that assembles the dlt resource is called only for airtables that the user desires to be loaded. The filter has already been done.
3. each pyairtable.Table is a quite lightweight dictionary. It only contains a little API request configuration but not the table's data. Thus, I am not sure if we gain much benefit by wrapping it into a generator.

But, based on your pointer I read up once again on the recent changes to pyairtable 2.0 and I refactored our code a bit. Thank you!

Unless you have further thoughts I would resolve this comment.